Rate limit technical design (#930)
kyma-project/istio: docs/contributor/adr/0007-rate-limit.md

# Rate Limit

## Status

Draft

## Context

It has been decided to indroduce rate limit functionality into Kyma platform to allow users to set rate limit functionality on the service mesh layer, therefore allowing to consume intended service mesh functionality which is abstracting away networking concerns outside applications inside the mesh.
Since Istio is an underlying service mesh responsible for the workloads networking across Kyma clusters, therefore it is the Istio which allows to configure such functionality.
Taking that into account Goat Team is responsible to introduce this feature.
At this moment it is not possible to introduce global rate limit provided by Istio service mesh because of technical or legal limitations related to the deployment of key-value stores in the cluster.

## Decision

> Taking the context part into account, it was decided to implement local rate limit as the only rate limiting posibility for now.

Reviewer: Does it make sense to explain here what does the local mean in this context (implicitly opposed to 'global')?

Author: Sure, it's a good idea since it can be confusing sometimes. I'll try to add some short explanation.

# Rate Limit

## Status

Draft

## Context

It has been decided to indroduce rate limit functionality into Kyma platform to allow users to set rate limit functionality on the service mesh layer, therefore allowing to consume intended service mesh functionality which is abstracting away networking concerns outside applications inside the mesh.
Since Istio is an underlying service mesh responsible for the workloads networking across Kyma clusters, therefore it is the Istio which allows to configure such functionality.
At this moment it is not possible to introduce global rate limit provided by Istio service mesh because of the technical or legal limitations, related to the deployment of key-value stores in the cluster.
Major difference between local and global rate limit is that local rate limiting is applied directly at the sidecar proxy level, controlling the rate of requests on a per-instance basis. Global rate limiting is applied accross multiple instances of a service or across, multiple services, providing a coorinated rate limit within the mesh.
In case of introduction the global rate limit functionality in the future, new ADR will be created.

> ## Decision

Reviewer 2: Did we think about enforcement: https://github.com/kyma-project/istio/blob/rate-limit-poc/docs/contributor/draft/rate-limiting-poc/01-rate-limiting.md#rate-limit-enforcement?

Reviewer 2: Envoy offers enforcement for 0 - 100% percentage of the traffic. I think the main usecase for enforcement disabled is, that you can check your rate limit metrics before blocking real traffic. Therefore I think it's easier for the user for to have a bool `enforced` to enable or disable enforcement. I think it would be good if this is a required configuration without any default, so the user must make a decision about this.

The following configuration is an example for a rate limit that is enabled, but not enforced:
```yaml
envoy.filters.http.local_ratelimit:
  "@type": type.googleapis.com/udpa.type.v1.TypedStruct
  type_url: type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
  value:
    stat_prefix: api_usage_local_rate_limiter
    enable_x_ratelimit_headers: DRAFT_VERSION_03
    filter_enabled:
      runtime_key: local_rate_limit_enabled
      default_value:
        numerator: 100
        denominator: HUNDRED
    filter_enforced:
      runtime_key: local_rate_limit_enforced
      default_value:
        numerator: 0
        denominator: HUNDRED
    token_bucket:
      max_tokens: 4
      tokens_per_fill: 2
      fill_interval: 60s
```

Author: Thanks for the work done here, I'll add it to the decision

Author: Also I'll do a modification to the API design

> ## Decision

Reviewer 2: Do we want to return the rate limit headers per default?
https://github.com/kyma-project/istio/blob/rate-limit-poc/docs/contributor/draft/rate-limiting-poc/01-rate-limiting.md#rate-limit-response-headers

Author: The only risk I can see in having those headers is that it gives a possibility to fine tune DoS attacks to the rate limit values to make detection harder. Second thing is potential insight into how many requests are getting to the system. I'd turn them off by default as the most secure option, and give users possibility to turn it on with one simple field.

Reviewer 2: The field would be a boolean `enableHeaders` on the level of `spec`?

Author: I think we can do enableHeaders. I'll add it to the decision

> It was decided to implement local rate limit as the only rate limiting option at this time.

Reviewer 2: I thought we wanted to mention that the global rate limiting is out of scope of this ADR, because of the limitations? 🤔  In this case I'd not mention it as a decision.

Author: Removed

> New CRD in the Istio module will be introduced to allow its configuration.

Reviewer 2: ```suggestion
A new CRD will be introduced in the Istio module to allow rate limiting configuration.
```

> It was decided to implement local rate limit as the only rate limiting option at this time.

Author: ```suggestion
```

In the future, global rate limit might be introduced. The decision has not been made yet, therefore it has been decided to name the new CRD `RateLimit` to not to indicate, that global rate limit will be provided. By naming current CRD for example `LocalRateLimit` logical would be to expect also `GlobalRateLimit` to be existent at some point.
Currently the possibilites for the rate limit functionality are limited to restrict number of requests, without any other criteria like header presence or header value. 
CRD is structured in a way that allows further extensions, allowing to configure different rate limit criteria, or to introduce global rate limit.

> ## Consequences

Reviewer 2: Maybe we should add a consequence that rate limited requests will result in a status code 429?

> ## Consequences

Reviewer 2: Do we want to have a consequence about documentation of rate limit metrics for user enablement?
https://github.com/kyma-project/istio/blob/rate-limit-poc/docs/contributor/draft/rate-limiting-poc/01-rate-limiting.md#rate-limit-metrics

Reviewer 2: Since the field `stat_prefix` for the local rate limit config in Envoy is required, we need to set a value. This value will be the prefix for the metrics. I think we don't need to make it configurable, yet. Also it might not be required to use a different prefix for each RateLimit configuration. 

My proposal is to use something simple for now for the local rate limiting like `stat_prefix: rate_limit`.
This will result in the following metrics:
```
rate_limit.http_local_rate_limit.enabled: 6
rate_limit.http_local_rate_limit.enforced: 0
rate_limit.http_local_rate_limit.ok: 4
rate_limit.http_local_rate_limit.rate_limited: 2
```

We should add this as a decision.

Author: thanks for handling that, I'm resolving it

New controller for the new rate limit CRD needs to be implemented as a part of the Istio module.
The new rate limit CR can be used by users to set rate limits in the service mesh in their cluster without having to worry about possible changes in the Istio EnvoyFilter resources.
Nevertheless EnvoyFilter is complex and since it's API is not stable yet, current RateLimit CR will be provided as v1alpha1 version, and might be changed in the future Istio releases.

> RateLimit CRD should be included into blocking deletion strategy for IstioCR, since Istio should not be uninstalled while RateLimit CRs exist in the cluster.

Reviewer 2: ```suggestion
RateLimit CRD should be included into blocking deletion strategy for Istio CR, since Istio should not be uninstalled while RateLimit CRs exist in the cluster.
```

New functionality requires integration tests to be implemented.

> In case of an introduction of any additional criteria to the local rate limit in the RateLimit CR, the capacity of rate limit on this criteria should be lower than or at most equal to the root setting. Otherwise it will be reduced to the root capacity value.

Reviewer 2: I think it might be hard to understand that. Maybe you can be a bit more descriptive and provide more context?

Reviewer 2: I just saw that this behaviour can be controlled with `always_consume_default_token_bucket`([docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto#local-rate-limit-proto)). So we should ignore it for now and completely remove it from the consequences. I think this is something to consider when we add more possibilities to local rate limiting.

```suggestion
```

Author: Sure, I'll try to describe it a bit clearer

Author: let's remove, the simpler, the better for now
